File: src/lib/farm-v2/engine.ts
// 农场 v1.2 纯函数引擎

import seedrandom from 'seedrandom';
import type {
  CropIdV2, FertilizerType, Quality, Season, WeatherV2, CropInstance,
  CropStageV2, FarmStateV2, ComputedLand, LandPlot, ProtectionBuffs,
} from '@/lib/types/farm-v2';
import {
  CROPS_V2, FERTILIZERS, NO_FERTILIZER_RATES, SEASON_MODIFIERS, WEATHERS_V2,
  SEASON_WEATHER_PROB, WATER_MISS_MUL, WATER_QUALITY_PENALTY, PERFECT_CARE_BONUS,
  OVERRIPE_TIERS, QUALITY_MULTIPLIERS, SEASON_EPOCH_DATE,
} from './config';

const CHINA_TZ_OFFSET_MS = 8 * 60 * 60 * 1000;
const WEEK_MS = 7 * 24 * 60 * 60 * 1000;

/** 获取中国时区时间字符串 YYYY-MM-DD */
export function getChinaDateString(ts: number): string {
  const d = new Date(ts + CHINA_TZ_OFFSET_MS);
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/** 中国时区今日 0 点的 UTC 时间戳 */
export function getChinaMidnight(ts: number): number {
  const day = Math.floor((ts + CHINA_TZ_OFFSET_MS) / (24 * 60 * 60 * 1000));
  return day * 24 * 60 * 60 * 1000 - CHINA_TZ_OFFSET_MS;
}

/** 计算某时刻所属的季节 */
export function getCurrentSeason(ts: number): Season {
  const epoch = parseEpochDate(SEASON_EPOCH_DATE);
  const seasonIndex = Math.floor((getChinaMidnight(ts) - epoch) / WEEK_MS);
  const arr: Season[] = ['spring', 'summer', 'autumn', 'winter'];
  const i = ((seasonIndex % 4) + 4) % 4;
  return arr[i];
}

/** 距离下次换季的毫秒 */
export function getNextSeasonChangeMs(now: number): number {
  const epoch = parseEpochDate(SEASON_EPOCH_DATE);
  const elapsed = getChinaMidnight(now) - epoch;
  const currentWeek = Math.floor(elapsed / WEEK_MS);
  const nextSeasonStart = epoch + (currentWeek + 1) * WEEK_MS;
  return Math.max(0, nextSeasonStart - now);
}

function parseEpochDate(s: string): number {
  const [y, m, d] = s.split('-').map(Number);
  // 中国时区 0 点
  return Date.UTC(y, m - 1, d) - CHINA_TZ_OFFSET_MS;
}

/** 距离下一次中国时区 0 点的毫秒数 */
export function getNextDailyResetMs(now: number): number {
  const today = getChinaMidnight(now);
  return Math.max(0, today + 24 * 60 * 60 * 1000 - now);
}

/** 计算实际成长时间（毫秒） */
export function computeActualGrowthMs(cropId: CropIdV2, season: Season, fert: FertilizerType): number {
  const crop = CROPS_V2[cropId];
  const seasonFactor = SEASON_MODIFIERS[season].growth;
  const fertFactor = fert ? FERTILIZERS[fert].growthFactor : 1;
  return Math.round(crop.growthMinutes * seasonFactor * fertFactor * 60 * 1000);
}

/** 计算实际浇水间隔（毫秒） */
export function computeActualWaterIntervalMs(cropId: CropIdV2, season: Season, weather: WeatherV2): number {
  const crop = CROPS_V2[cropId];
  const seasonFactor = SEASON_MODIFIERS[season].water;
  const weatherFactor = WEATHERS_V2[weather].waterFactor;
  return Math.round(crop.waterIntervalMinutes * seasonFactor * weatherFactor * 60 * 1000);
}

/** 计算作物当前阶段（基于成长进度） */
export function computeCropStage(progress: number): CropStageV2 {
  if (progress >= 1) return 'mature';
  if (progress >= 0.5) return 'growing';
  if (progress >= 0.2) return 'sprout';
  return 'seed';
}

/** 计算成长进度 0..1 */
export function computeGrowthProgress(crop: CropInstance, now: number): number {
  if (now >= crop.matureAt) return 1;
  const total = crop.matureAt - crop.plantedAt;
  if (total <= 0) return 1;
  return Math.max(0, (now - crop.plantedAt) / total);
}

/** 计算缺水次数变化（懒结算） */
export function computeWaterMissesAfterWindow(
  crop: CropInstance,
  intervalMs: number,
  now: number,
): { newMissCount: number; newNextDue: number; newLastWater: number } {
  // 仅未成熟期间累积缺水
  const checkUntil = Math.min(now, crop.matureAt);
  let nextDue = crop.nextWaterDueAt;
  let missCount = crop.waterMissCount;
  while (checkUntil > nextDue && missCount < 3) {
    missCount += 1;
    nextDue += intervalMs;
  }
  return { newMissCount: missCount, newNextDue: nextDue, newLastWater: crop.lastWaterAt };
}

/** 缺水收益系数 */
export function getWaterPenaltyMul(missCount: number): number {
  if (missCount >= 3) return 0;
  return WATER_MISS_MUL[missCount] ?? 0;
}

/** 过熟系数 */
export function computeOverripeFactor(crop: CropInstance, now: number): number {
  if (now < crop.matureAt) return 1;
  const hours = (now - crop.matureAt) / (60 * 60 * 1000);
  for (const [tierH, factor] of OVERRIPE_TIERS) {
    if (hours <= tierH) return factor;
  }
  return 0;
}

/** 完美照顾判定 */
export function isPerfectCare(crop: CropInstance, now: number): boolean {
  return crop.waterMissCount === 0
    && crop.stolenCount === 0
    && now < crop.matureAt + 12 * 60 * 60 * 1000;
}

/** 品质概率：返回归一化后的 [普通, 银, 金] */
export function rollQualityRates(
  fert: FertilizerType,
  missCount: number,
  perfect: boolean,
): [number, number, number] {
  const base = fert ? FERTILIZERS[fert].qualityRates : NO_FERTILIZER_RATES;
  let [n, s, g] = base;

  // 缺水惩罚（金/银 乘数）
  if (missCount > 0 && missCount <= 2) {
    const [goldMul, silverMul] = WATER_QUALITY_PENALTY[missCount];
    g = g * goldMul;
    s = s * silverMul;
  } else if (missCount >= 3) {
    return [1, 0, 0];
  }

  // 完美照顾
  if (perfect) {
    s += PERFECT_CARE_BONUS[0];
    g += PERFECT_CARE_BONUS[1];
  }

  // 归一化
  s = Math.max(0, s);
  g = Math.max(0, g);
  n = Math.max(0, 1 - s - g);
  const total = n + s + g;
  if (total <= 0) return [1, 0, 0];
  return [n / total, s / total, g / total];
}

/** 根据概率随机品质 */
export function pickQuality(rates: [number, number, number], rng: () => number): Quality {
  const r = rng();
  if (r < rates[0]) return 'normal';
  if (r < rates[0] + rates[1]) return 'silver';
  return 'gold';
}

/** 计算最终收获积分 */
export function computeFinalYield(
  cropId: CropIdV2,
  quality: Quality,
  missCount: number,
  season: Season,
  overripe: number,
  stolenAmount: number,
): number {
  const base = CROPS_V2[cropId].baseYield;
  const qm = QUALITY_MULTIPLIERS[quality];
  const wm = getWaterPenaltyMul(missCount);
  const sm = SEASON_MODIFIERS[season].yield;
  const raw = base * qm * wm * sm * overripe;
  return Math.max(0, Math.floor(raw) - stolenAmount);
}

/** 加权抽取天气（按 SEASON_WEATHER_PROB 表） */
export function getWeatherForDate(dateStr: string, season: Season): WeatherV2 {
  const rng = seedrandom(`weather:${dateStr}:${season}`);
  const r = rng();
  const dist = SEASON_WEATHER_PROB[season];
  let acc = 0;
  for (const [w, p] of dist) {
    acc += p;
    if (r < acc) return w;
  }
  return dist[0][0];
}

/** 稻草人是否生效 */
export function isScarecrowActive(state: FarmStateV2, now: number): boolean {
  return state.scarecrowUntil != null && state.scarecrowUntil > now;
}

export function isBellActive(state: FarmStateV2, now: number): boolean {
  return state.bellUntil != null && state.bellUntil > now;
}

/** 自家宠物是否守护中 */
export function isPetGuarding(state: FarmStateV2, now: number): boolean {
  const pet = state.pet;
  return !!pet && pet.currentTask === 'guard'
    && pet.taskEndAt != null && pet.taskEndAt > now;
}

/** 自家宠物是否在赶乌鸦 */
export function isPetChasing(state: FarmStateV2, now: number): boolean {
  const pet = state.pet;
  return !!pet && pet.currentTask === 'chase_crow'
    && pet.taskEndAt != null && pet.taskEndAt > now;
}

/** 计算预估品质提示（用于 UI） */
export function estimateQualityHint(crop: CropInstance, now: number): Quality | null {
  if (now < crop.matureAt) return null;
  const perfect = isPerfectCare(crop, now);
  const rates = rollQualityRates(crop.fertilizer, crop.waterMissCount, perfect);
  if (rates[2] >= 0.3) return 'gold';
  if (rates[1] >= 0.3) return 'silver';
  return 'normal';
}

/** 推导土地状态（仅展示用，不修改原 LandPlot） */
export function computeLandStatus(
  plot: LandPlot,
  now: number,
): { status: import('@/lib/types/farm-v2').LandStatus; stage: CropStageV2 | null; progress: number; remaining: number; nextWater: number } {
  if (plot.status === 'locked') {
    return { status: 'locked', stage: null, progress: 0, remaining: 0, nextWater: 0 };
  }
  if (plot.status === 'withered') {
    return { status: 'withered', stage: null, progress: 0, remaining: 0, nextWater: 0 };
  }
  if (plot.status === 'eaten') {
    return { status: 'eaten', stage: null, progress: 0, remaining: 0, nextWater: 0 };
  }
  if (!plot.crop) {
    return { status: 'empty', stage: null, progress: 0, remaining: 0, nextWater: 0 };
  }
  const crop = plot.crop;

  // 缺水 3+ 应当已经在 tick 中变为 withered；这里防御性检查
  if (crop.waterMissCount >= 3) {
    return { status: 'withered', stage: null, progress: 0, remaining: 0, nextWater: 0 };
  }
  const progress = computeGrowthProgress(crop, now);
  const stage = computeCropStage(progress);
  const remaining = Math.max(0, crop.matureAt - now);
  const nextWater = stage === 'mature' ? 0 : Math.max(0, crop.nextWaterDueAt - now);

  if (stage === 'mature') {
    return { status: 'mature', stage, progress, remaining: 0, nextWater: 0 };
  }
  if (plot.status === 'thirsty' && crop.waterMissCount > 0) {
    return { status: 'thirsty', stage, progress, remaining, nextWater };
  }
  // 当前是否正在缺水（已超过 nextWaterDueAt 但还未触发新的 miss）
  if (now > crop.nextWaterDueAt) {
    return { status: 'thirsty', stage, progress, remaining, nextWater: 0 };
  }
  return { status: 'growing', stage, progress, remaining, nextWater };
}

/** 构建 ComputedLand 数组 */
export function buildComputedLands(
  state: FarmStateV2,
  now: number,
): ComputedLand[] {
  return state.lands.map((plot) => {
    const cs = computeLandStatus(plot, now);
    const overripe = plot.crop ? computeOverripeFactor(plot.crop, now) : 1;
    const hint = plot.crop ? estimateQualityHint(plot.crop, now) : null;
    return {
      ...plot,
      status: cs.status,
      stage: cs.stage,
      growthProgress: cs.progress,
      remainingMs: cs.remaining,
      nextWaterRemainingMs: cs.nextWater,
      overripeFactor: overripe,
      expectedQualityHint: hint,
      scarecrowActive: isScarecrowActive(state, now),
      bellActive: isBellActive(state, now),
      netActive: !!(plot.crop?.birdNetUntil && plot.crop.birdNetUntil > now),
    };
  });
}

/** 当前可种作物：必须解锁（landIndex >= unlockLandCount）且季节匹配 */
export function getPlantableCrops(state: FarmStateV2, season: Season): CropIdV2[] {
  const unlockedLandCount = state.lands.filter((l) => l.status !== 'locked').length;
  const ids: CropIdV2[] = [];
  for (const cid of Object.keys(CROPS_V2) as CropIdV2[]) {
    const def = CROPS_V2[cid];
    if (def.unlockLandCount > unlockedLandCount) continue;
    if (!def.seasons.includes(season)) continue;
    ids.push(cid);
  }
  return ids;
}

/** 拼装防护状态 */
export function buildProtections(state: FarmStateV2, now: number): ProtectionBuffs {
  return {
    scarecrowActive: isScarecrowActive(state, now),
    bellActive: isBellActive(state, now),
    petGuarding: isPetGuarding(state, now),
    petChasing: isPetChasing(state, now),
  };
}

File: backend/internal/farm/engine.go
package farm

import (
	"fmt"
	"math"
	"time"
)

const (
	chinaTZOffsetMs = int64(8 * 60 * 60 * 1000)
	dayMs           = int64(24 * 60 * 60 * 1000)
	weekMs          = int64(7 * 24 * 60 * 60 * 1000)
	seasonEpochDate = "2025-01-05"
)

var seasonOrder = []Season{SeasonSpring, SeasonSummer, SeasonAutumn, SeasonWinter}

type cropDef struct {
	ID              CropID
	Name            string
	Seasons         []Season
	SeedCost        int64
	GrowthMinutes   int64
	WaterMinutes    int64
	BaseYield       int64
	UnlockLandCount int
}

var cropDefs = []cropDef{
	{ID: CropWheat, Name: "小麦", Seasons: []Season{SeasonSpring, SeasonAutumn}, SeedCost: 5, GrowthMinutes: 30, WaterMinutes: 30, BaseYield: 12, UnlockLandCount: 4},
	{ID: CropCarrot, Name: "胡萝卜", Seasons: []Season{SeasonSpring, SeasonAutumn}, SeedCost: 8, GrowthMinutes: 60, WaterMinutes: 30, BaseYield: 20, UnlockLandCount: 4},
	{ID: CropLettuce, Name: "生菜", Seasons: []Season{SeasonSpring}, SeedCost: 10, GrowthMinutes: 90, WaterMinutes: 30, BaseYield: 28, UnlockLandCount: 4},
	{ID: CropTomato, Name: "番茄", Seasons: []Season{SeasonSummer}, SeedCost: 18, GrowthMinutes: 120, WaterMinutes: 40, BaseYield: 48, UnlockLandCount: 5},
	{ID: CropPotato, Name: "土豆", Seasons: []Season{SeasonWinter, SeasonSpring}, SeedCost: 20, GrowthMinutes: 150, WaterMinutes: 60, BaseYield: 55, UnlockLandCount: 5},
	{ID: CropStrawberry, Name: "草莓", Seasons: []Season{SeasonSpring, SeasonSummer}, SeedCost: 25, GrowthMinutes: 180, WaterMinutes: 45, BaseYield: 75, UnlockLandCount: 6},
	{ID: CropCorn, Name: "玉米", Seasons: []Season{SeasonSummer, SeasonAutumn}, SeedCost: 35, GrowthMinutes: 240, WaterMinutes: 60, BaseYield: 105, UnlockLandCount: 7},
	{ID: CropPumpkin, Name: "南瓜", Seasons: []Season{SeasonAutumn}, SeedCost: 45, GrowthMinutes: 360, WaterMinutes: 90, BaseYield: 150, UnlockLandCount: 8},
}

var landUnlockPrices = map[int]int64{
	1: 0,
	2: 0,
	3: 0,
	4: 0,
	5: 50,
	6: 100,
	7: 150,
	8: 200,
}

var cropDefByID = func() map[CropID]cropDef {
	result := map[CropID]cropDef{}
	for _, crop := range cropDefs {
		result[crop.ID] = crop
	}
	return result
}()

var seasonWaterFactor = map[Season]float64{
	SeasonSpring: 1.00,
	SeasonSummer: 0.85,
	SeasonAutumn: 1.00,
	SeasonWinter: 1.20,
}

var seasonGrowthFactor = map[Season]float64{
	SeasonSpring: 0.95,
	SeasonSummer: 0.90,
	SeasonAutumn: 1.00,
	SeasonWinter: 1.15,
}

var weatherWaterFactor = map[Weather]float64{
	WeatherSunny:     1.00,
	WeatherCloudy:    1.10,
	WeatherLightRain: 1.00,
	WeatherStorm:     1.00,
	WeatherHot:       0.80,
	WeatherWind:      1.00,
	WeatherSnow:      1.20,
	WeatherFog:       1.00,
}

var weatherAutoWaterMinutes = map[Weather]int64{
	WeatherSunny:     0,
	WeatherCloudy:    0,
	WeatherLightRain: 30,
	WeatherStorm:     15,
	WeatherHot:       0,
	WeatherWind:      0,
	WeatherSnow:      0,
	WeatherFog:       0,
}

var weatherCrowFactor = map[Weather]float64{
	WeatherSunny:     1.00,
	WeatherCloudy:    0.90,
	WeatherLightRain: 0.40,
	WeatherStorm:     0.00,
	WeatherHot:       1.20,
	WeatherWind:      1.50,
	WeatherSnow:      0.30,
	WeatherFog:       0.70,
}

var seasonCrowFactor = map[Season]float64{
	SeasonSpring: 1.00,
	SeasonSummer: 1.20,
	SeasonAutumn: 1.00,
	SeasonWinter: 0.70,
}

var seasonWeatherProb = map[Season][]struct {
	weather Weather
	prob    float64
}{
	SeasonSpring: {
		{WeatherSunny, 0.30}, {WeatherCloudy, 0.25}, {WeatherLightRain, 0.30}, {WeatherStorm, 0.05}, {WeatherWind, 0.10},
	},
	SeasonSummer: {
		{WeatherSunny, 0.25}, {WeatherCloudy, 0.10}, {WeatherLightRain, 0.20}, {WeatherStorm, 0.15}, {WeatherHot, 0.30},
	},
	SeasonAutumn: {
		{WeatherSunny, 0.40}, {WeatherCloudy, 0.25}, {WeatherLightRain, 0.10}, {WeatherWind, 0.20}, {WeatherFog, 0.05},
	},
	SeasonWinter: {
		{WeatherSunny, 0.25}, {WeatherCloudy, 0.25}, {WeatherWind, 0.15}, {WeatherSnow, 0.30}, {WeatherFog, 0.05},
	},
}

func getChinaDateString(ts int64) string {
	d := time.UnixMilli(ts + chinaTZOffsetMs).UTC()
	return d.Format("2006-01-02")
}

func getChinaMidnight(ts int64) int64 {
	day := floorDiv(ts+chinaTZOffsetMs, dayMs)
	return day*dayMs - chinaTZOffsetMs
}

func getCurrentSeason(ts int64) Season {
	epoch := parseEpochDate(seasonEpochDate)
	seasonIndex := floorDiv(getChinaMidnight(ts)-epoch, weekMs)
	i := int(((seasonIndex % 4) + 4) % 4)
	return seasonOrder[i]
}

func getNextSeasonChangeMs(now int64) int64 {
	epoch := parseEpochDate(seasonEpochDate)
	elapsed := getChinaMidnight(now) - epoch
	currentWeek := floorDiv(elapsed, weekMs)
	nextSeasonStart := epoch + (currentWeek+1)*weekMs
	return maxInt64(0, nextSeasonStart-now)
}

func getNextDailyResetMs(now int64) int64 {
	today := getChinaMidnight(now)
	return maxInt64(0, today+dayMs-now)
}

func parseEpochDate(value string) int64 {
	t, err := time.Parse("2006-01-02", value)
	if err != nil {
		panic(fmt.Sprintf("invalid farm season epoch %q: %v", value, err))
	}
	return t.UnixMilli() - chinaTZOffsetMs
}

func getWeatherForDate(dateStr string, season Season) Weather {
	rng := newSeedRandom(fmt.Sprintf("weather:%s:%s", dateStr, season))
	r := rng.Float64()
	dist := seasonWeatherProb[season]
	if len(dist) == 0 {
		dist = seasonWeatherProb[SeasonSpring]
	}
	acc := 0.0
	for _, item := range dist {
		acc += item.prob
		if r < acc {
			return item.weather
		}
	}
	return dist[0].weather
}

func computeCropStage(progress float64) CropStage {
	if progress >= 1 {
		return CropStageMature
	}
	if progress >= 0.5 {
		return CropStageGrowing
	}
	if progress >= 0.2 {
		return CropStageSprout
	}
	return CropStageSeed
}

func computeGrowthProgress(crop CropInstance, now int64) float64 {
	if now >= crop.MatureAt {
		return 1
	}
	total := crop.MatureAt - crop.PlantedAt
	if total <= 0 {
		return 1
	}
	return math.Max(0, float64(now-crop.PlantedAt)/float64(total))
}

func computeOverripeFactor(crop CropInstance, now int64) float64 {
	if now < crop.MatureAt {
		return 1
	}
	hours := float64(now-crop.MatureAt) / float64(60*60*1000)
	switch {
	case hours <= 12:
		return 1
	case hours <= 24:
		return 0.8
	case hours <= 48:
		return 0.5
	default:
		return 0
	}
}

func computeActualWaterIntervalMs(cropID CropID, season Season, weather Weather) int64 {
	crop, ok := cropDefByID[cropID]
	if !ok {
		return 30 * 60 * 1000
	}
	seasonFactor := seasonWaterFactor[season]
	if seasonFactor == 0 {
		seasonFactor = 1
	}
	weatherFactor := weatherWaterFactor[weather]
	if weatherFactor == 0 {
		weatherFactor = 1
	}
	return int64(math.Round(float64(crop.WaterMinutes) * seasonFactor * weatherFactor * 60 * 1000))
}

func computeActualGrowthMs(cropID CropID, season Season) int64 {
	crop, ok := cropDefByID[cropID]
	if !ok {
		return 30 * 60 * 1000
	}
	seasonFactor := seasonGrowthFactor[season]
	if seasonFactor == 0 {
		seasonFactor = 1
	}
	return int64(math.Round(float64(crop.GrowthMinutes) * seasonFactor * 60 * 1000))
}

func computeWaterMissesAfterWindow(crop CropInstance, intervalMs int64, now int64) (int64, int64) {
	checkUntil := now
	if crop.MatureAt < checkUntil {
		checkUntil = crop.MatureAt
	}
	nextDue := crop.NextWaterDueAt
	missCount := crop.WaterMissCount
	for checkUntil > nextDue && missCount < 3 {
		missCount++
		nextDue += intervalMs
	}
	return missCount, nextDue
}

func isPerfectCare(crop CropInstance, now int64) bool {
	return crop.WaterMissCount == 0 && crop.StolenCount == 0 && now < crop.MatureAt+12*60*60*1000
}

func estimateQualityHint(crop CropInstance, now int64) *Quality {
	if now < crop.MatureAt {
		return nil
	}
	quality := QualityNormal
	if isPerfectCare(crop, now) {
		quality = QualitySilver
	}
	return &quality
}

func buildComputedLands(state FarmState, now int64) []ComputedLand {
	lands := make([]ComputedLand, 0, len(state.Lands))
	scarecrowActive := state.ScarecrowUntil != nil && *state.ScarecrowUntil > now
	bellActive := state.BellUntil != nil && *state.BellUntil > now
	for _, plot := range state.Lands {
		computed := ComputedLand{
			Index:                plot.Index,
			Status:               plot.Status,
			Crop:                 plot.Crop,
			Stage:                nil,
			GrowthProgress:       0,
			RemainingMs:          0,
			NextWaterRemainingMs: 0,
			OverripeFactor:       1,
			ExpectedQualityHint:  nil,
			ScarecrowActive:      scarecrowActive,
			BellActive:           bellActive,
			NetActive:            plot.Crop != nil && plot.Crop.BirdNetUntil != nil && *plot.Crop.BirdNetUntil > now,
		}

		switch {
		case plot.Status == LandStatusLocked:
			computed.Status = LandStatusLocked
		case plot.Status == LandStatusWithered:
			computed.Status = LandStatusWithered
		case plot.Status == LandStatusEaten:
			computed.Status = LandStatusEaten
		case plot.Crop == nil:
			computed.Status = LandStatusEmpty
		default:
			crop := *plot.Crop
			if crop.WaterMissCount >= 3 {
				computed.Status = LandStatusWithered
				computed.Crop = nil
				break
			}
			progress := computeGrowthProgress(crop, now)
			stage := computeCropStage(progress)
			computed.Stage = &stage
			computed.GrowthProgress = progress
			computed.RemainingMs = maxInt64(0, crop.MatureAt-now)
			computed.OverripeFactor = computeOverripeFactor(crop, now)
			computed.ExpectedQualityHint = estimateQualityHint(crop, now)
			if stage == CropStageMature {
				computed.Status = LandStatusMature
				computed.RemainingMs = 0
				computed.NextWaterRemainingMs = 0
			} else if plot.Status == LandStatusThirsty && crop.WaterMissCount > 0 {
				computed.Status = LandStatusThirsty
				computed.NextWaterRemainingMs = maxInt64(0, crop.NextWaterDueAt-now)
			} else if now > crop.NextWaterDueAt {
				computed.Status = LandStatusThirsty
				computed.NextWaterRemainingMs = 0
			} else {
				computed.Status = LandStatusGrowing
				computed.NextWaterRemainingMs = maxInt64(0, crop.NextWaterDueAt-now)
			}
		}

		lands = append(lands, computed)
	}
	return lands
}

func getPlantableCrops(state FarmState, season Season) []CropID {
	unlockedLandCount := 0
	for _, land := range state.Lands {
		if land.Status != LandStatusLocked {
			unlockedLandCount++
		}
	}
	ids := []CropID{}
	for _, crop := range cropDefs {
		if crop.UnlockLandCount > unlockedLandCount {
			continue
		}
		if !seasonContains(crop.Seasons, season) {
			continue
		}
		ids = append(ids, crop.ID)
	}
	return ids
}

func seasonContains(seasons []Season, season Season) bool {
	for _, item := range seasons {
		if item == season {
			return true
		}
	}
	return false
}

func floorDiv(a, b int64) int64 {
	q := a / b
	r := a % b
	if r != 0 && ((r < 0) != (b < 0)) {
		q--
	}
	return q
}

func maxInt64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func minInt64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

var timeFromUnixMilli = func(ms int64) time.Time {
	return time.UnixMilli(ms)
}
